core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -94,8 +88,20 @@ class LogSegment private[log] (val log: FileRecords,
   private var rollingBasedTimestamp: Option[Long] = None
 
   /* The maximum timestamp we see so far */
-  @volatile private var maxTimestampSoFar: Long = timeIndex.lastEntry.timestamp
-  @volatile private var offsetOfMaxTimestamp: Long = timeIndex.lastEntry.offset
+  @volatile private var _maxTimestampSoFar: Long = -1L
+  def maxTimestampSoFar: Long = {
+    if (_maxTimestampSoFar < 0) {
+      _maxTimestampSoFar = timeIndex.get.lastEntry.timestamp
+    }
+    _maxTimestampSoFar
+  }
+  @volatile private var _offsetOfMaxTimestamp: Long = -1L
+  def offsetOfMaxTimestamp: Long = {
+    if (_offsetOfMaxTimestamp < 0) {
+      _offsetOfMaxTimestamp = timeIndex.get.lastEntry.offset
+    }
+    _offsetOfMaxTimestamp
```

Comment:
Why not handle this in the `Getter` class?

Reply:
The reason is that these fields will be updated outside of the index (e.g. https://github.com/apache/kafka/pull/5498/files/44c2d970169f2b85c26af50ecfab4796802abb1a#diff-ac167eb8ca7cc10b6b8fa5772fa438f6R148). If we always access the index to get back these fields, the value can be stale because the index will only get updated when `bytesSinceLastIndexEntry > indexIntervalBytes`. Even if the stale value is fine (but this is a correctness issue),  we don't want to re-read the last entry of the index after the append just in order to get back what we have already known.